### DISCUSSION TITLE
Basic question about properties and viewmodels

### DISCUSSION THREAD

**Discussion by sonicsmooth** [AvaloniaUI/Avalonia#9815] [General]
Avalonia newbie still.  Specific questions at the bottom.

I have a custom control which draws stuff in `Render()` using properties defined in codebehind, and specified in xaml.

GraphGrid.xaml.cs is codebehind

    namespace avgrid.Views;
    public partial class GraphGrid : Panel
    {
        public static readonly DirectProperty<GraphGrid, Double> ScaleProperty = 
            AvaloniaProperty.RegisterDirect<GraphGrid, Double>(nameof(Scale),  o => o.Scale,  (o,v) => o.Scale = v);
        private Double _scale = 2.5;
        public Double Scale
        {
            get => _scale;
            set => this.SetAndRaise(ScaleProperty, ref _scale, value);
        }
        ...
        public override void Render(DrawingContext dc) { ... }
    }

Nothing in GraphGrid.xaml

    <Panel xmlns="https://github.com/avaloniaui"
                xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
                xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
                xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
                mc:Ignorable="d" d:DesignWidth="800" d:DesignHeight="450"
                x:Class="avgrid.Views.GraphGrid"/>
    
GraphGridViewModel.cs has the same property, repeated

    namespace avgrid.ViewModels;
    public class GraphGridViewModel : ViewModelBase
    {
        private Double _scale = 1.25;
        public Double Scale 
        { 
            get => _scale;
            set => this.RaiseAndSetIfChanged(ref _scale, value);
        }
        public void Reset()
        {
            Console.WriteLine("GraphGrid ViewModel Reset");
            Scale = 1.0;
        }
    }

Custom control instantiated in MainWindow.xaml

    <Window xmlns="https://github.com/avaloniaui"
            xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
            xmlns:vm="using:avgrid.ViewModels"
            xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
            xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
            xmlns:controls="clr-namespace:avgrid.Views"
            mc:Ignorable="d" d:DesignWidth="600" d:DesignHeight="400"
            Width="600" Height="400"
            x:Class="avgrid.Views.MainWindow"
            Icon="/Assets/avalonia-logo.ico"
            Title="avgrid">

      <Grid ...>
        <Grid.DataContext>
          <vm:GraphGridViewModel/>
        </Grid.DataContext>

        <controls:GraphGrid
          Scale="{Binding Scale}"> 
        </controls:GraphGrid>
        <StackPanel ...>
          <Button Command="{Binding Reset}" Content="Reset"/>
        </StackPanel>
      </Grid>
    </Window>

Questions:
1.  Is it correct and common to have a property defined twice -- once in the codebehind for usage in xaml, and once in the viewmodel for binding to?  Specifically, is `<GraphGrid Scale="{Binding Scale}">` idiomatic?  It seems to work in practice, but it seems I'm missing something simpler.
2. In the `MainWindow`, I set the `Grid.Datacontext` to the `GraphGridViewModel` so the `{Binding Scale}` and the `{Binding Reset}` work correctly rather than trying to bind to something in the `MainWindowViewModel`.  This works, but again I'm not sure if this is how things are normally done.  I can instantiate the `GraphGridViewModel` as datacontext in `GraphGrid.xaml` which seems correct, but then how do I access that instance so the `Reset` binding works?  I've also seen examples (Music Store example) where the inner viewmodel is created in the MainWindowViewModel, but then I'm not sure how to ensure the bindings work properly.

Thanks for your patience with newb questions.

**Comment by robloo** [comment]
> Is it correct and common to have a property defined twice

It's not overly common in custom controls, but yes, this happens -- more often with models/viewmodels.

Firstly you have to decide if you are creating a "view" or a "control". A "view" which is just any page or subsection of your app isn't a re-usable component quite like a control is. I'm assuming you are creating a control here.

Usually for controls there is NO view model. By that I mean the "code-behind" .xaml.cs contains the view model implementation (usually just properties and logic) and is itself the context for binding.

In your situation it might make more sense to create a `TemplatedControl` then you can add the properties to the .cs file and then reference them in the control template with `TemplatedBinding`. This is usually what we do with custom controls and would normally make sense for a graph-type control.

This control would then be used anywhere -- including a window -- but wouldn't have the Window within itself. 

You might want to look at the `Carousel` control https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Controls/Carousel.cs and https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Themes.Fluent/Controls/Carousel.xaml. This is a simple example and shows you properties as well as binding to methods `Previous` and `Next`.

**Reply by sonicsmooth** [reply]
Thanks, I will check the link.  Yes this GraphGrid is a control to be used in a few places.  Currently all the logic resides in the codebehind, but because it needs to be able to accept commands (eg Reset(), Pan(), Zoom()) from the outside world as well as from direct user interaction, I therefore thought I needed a ViewModel as a place where outside code could call those commands.

I'll keep playing around with where to instantiate the datacontext and then how to reference it from different places in the xaml until I can pose a well formed question.

**Reply by workgroupengineering** [reply]
You can using same strategy of xtc CameraView ShutterCommand https://github.com/xamarin/XamarinCommunityToolkit/blob/main/src/CommunityToolkit/Xamarin.CommunityToolkit/Views/CameraView/CameraView.shared.cs#L46

**Reply by sonicsmooth** [reply]
Yes I was thinking events for this thing rather than functions, maybe.  I need to read up on the the pros and cons for this new UI framework.
